### PR TITLE
Admin, List of Orders & Orders Bulk Management pages: use a date range picker

### LINKED ISSUE #6730: As an enterprise manager, I can use a date range picker
## Description

**- As a:** Enterprise user and super admin

**- On page:** 
 - `/admin/reports` on each reports filters
 - `/admin/orders?q[s]=completed_at+desc` in start / end date filters
 - `/admin/orders/bulk_management` in start / end date filters
 - `/admin/subscriptions/` when creating or editing a new subscription

**- I want to be able to do:** 
Have a date(time)-range-picker to select start/end date using our new date picker (thanks to #6664). 
See example here: https://flatpickr.js.org/examples/#range-calendar

## Acceptance Criteria & Tests
_Scenario 1 As a simple producer_
We need to make sure selecting date ranges ends up with correct results on all the following pages:
 - `/admin/reports` on each reports filters
 - `/admin/orders?q[s]=completed_at+desc` in start / end date filters
 - `/admin/orders/bulk_management` in start / end date filters
 - `/admin/subscriptions/` when creating or editing a new subscription

_Scenario 2 As a multi-producer hub_
Repeat same tests as scenario 1

_Scenario 3 As a super admin_
Repeat same tests as scenario 2

## Previous discussions
See https://github.com/openfoodfoundation/openfoodnetwork/issues/6664#issuecomment-760964353

### PR DESCRIPTION
**Pull request by jibees**
#### What? Why?

Partially closes #6730

This one introduce two things: 
 - We now use stimulus (and https://github.com/adrienpoly/stimulus-flatpickr) for the date(time)picker. It's simpler, more reliable, and we can update easily through `yarn`. It's the move we try to follow in global (ie. using `stimulus` for javascript DOM operation and `yarn` for packages management)
 - We now use the date range picker from flatpickr for: 
   - `/admin/orders` and `/admin/orders/bulk_management`
   - I decided to not use for `/admin/reports` and `/admin/subscriptions/` because it needs to select also time, and it's not that easy in term of UX. 

Capture d’écran 2022-08-08 à 13 55 24

#### What should we test?
- As I refactored each date(time)picker, we should test each ones, ie.:
  - `/admin/order_cycles` on ordercycle list but also when creating / editing an order cycle
  - `/admin/reports` on each reports filters
  - `/admin/subscriptions/` when creating or editing a new subscription
- Also, test date range picker from `/admin/orders` and `/admin/orders/bulk_management` pages.

#### Release notes

Changelog Category: User facing changes

### CODE REVIEW

File: yarn.lock [hunk 1]
```diff
@@ -2139,23 +2139,6 @@
...
-    "@storybook/client-logger" "6.5.9"
-    "@storybook/core-events" "6.5.9"
-    "@storybook/csf" "0.0.2--canary.4566f4d.1"
-    "@storybook/router" "6.5.9"
-    "@storybook/theming" "6.5.9"
-    "@types/webpack-env" "^1.16.0"
-    core-js "^3.8.2"
-    global "^4.4.0"
-    regenerator-runtime "^0.13.7"
-
```
Comment (mkllnk) [inline yarn.lock:2158]: This is an unrelated stuff-up by dependabot, right? Shall we exclude Storybook from Dependabot?

File: app/webpacker/packs/admin.js [hunk 2]
```diff
@@ -4,3 +4,6 @@ import { definitionsFromContext } from "stimulus/webpack-helpers";
 const application = Application.start();
 const context = require.context("controllers", true, /.js$/);
 application.load(definitionsFromContext(context));
+
+// Import style for flatpickr
+require("flatpickr/dist/flatpickr.css");
```
Comment (mkllnk) [inline app/webpacker/packs/admin.js:9]: Why here? What happened to the JS? The commit message doesn't explain anything.

File: app/webpacker/controllers/flatpickr_controller.js [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+      buttons.unshift({
+        label: datetimepicker
+          ? Spree.translations.now
+          : Spree.translations.today,
+      });
+    }
+    return [
+      ShortcutButtonsPlugin({
+        button: buttons,
+        label: mode != "range" && "or",
```
Comment (mkllnk) [inline app/webpacker/controllers/flatpickr_controller.js:88]: My JS knowledge is so outdated, I don't even know what this expression does.

File: app/webpacker/controllers/flatpickr_controller.js [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+import { de } from "flatpickr/dist/l10n/de";
+import { fr } from "flatpickr/dist/l10n/fr";
+import { it } from "flatpickr/dist/l10n/it";
+import { nl } from "flatpickr/dist/l10n/nl";
+import { pl } from "flatpickr/dist/l10n/pl";
+import { pt } from "flatpickr/dist/l10n/pt";
+import { ru } from "flatpickr/dist/l10n/ru";
+import { sv } from "flatpickr/dist/l10n/sv";
+import { tr } from "flatpickr/dist/l10n/tr";
+import { en } from "flatpickr/dist/l10n/default.js";
```
Comment (mkllnk) [inline app/webpacker/controllers/flatpickr_controller.js:15]: I bet that we forget this when adding a new language. Wouldn't it be better to use `I18n.available_locales` here?

**Review by mkllnk — COMMENTED** [review]
Nice! :clap:

File: yarn.lock (hunk 1 above)
Comment (jibees) [reply yarn.lock:2158]: Yep, error with `yarn.lock` on master, will rebase this PR.

File: app/webpacker/controllers/flatpickr_controller.js (hunk 3 above)
Comment (jibees) [reply app/webpacker/controllers/flatpickr_controller.js:88]:
```
if mode != "range"
  label = "or"
else
  label = undefined
end
```

Actually this evaluate the expression and return the value of the last element of the logical expression

File: app/webpacker/packs/admin.js (hunk 2 above)
Comment (jibees) [reply app/webpacker/packs/admin.js:9]: You're right, that should not be here. I followed the tutorial but it should be in `app/webpacker/css/admin/all.scss` (relatively on top to avoid CSS errors due to order)

File: app/webpacker/controllers/flatpickr_controller.js (hunk 4 above)
Comment (jibees) [reply app/webpacker/controllers/flatpickr_controller.js:15]: Yes, I'm pretty sure too. But adding a new language on OFN doesn't mean it exists on flatpickr. So we might have some trouble using automatic logic...

File: app/webpacker/controllers/flatpickr_controller.js (hunk 3 above)
Comment (mkllnk) [reply app/webpacker/controllers/flatpickr_controller.js:88]: Okay, that's actually what I thought but it didn't make much sense to me. What is the label `or` doing and why `undefined` otherwise?

Comment (mkllnk) [reply app/webpacker/controllers/flatpickr_controller.js:88]: Now I read the documentation of the buttons plugin and the code a bit more and I understand the *why*. I'm not sure we need the "or" and if we need it then we should translate it.

But I also noticed a lot of if-else branches around what kind of data we are selecting and I'm wondering if it would be better to store each configuration in a separate object and then apply that generically.

File: app/webpacker/controllers/flatpickr_controller.js (hunk 4 above)
Comment (mkllnk) [reply app/webpacker/controllers/flatpickr_controller.js:15]: In that case we need a list of available flatpickr translations as well and then we select from that.

File: app/webpacker/controllers/flatpickr_controller.js (hunk 3 above)
Comment (jibees) [reply app/webpacker/controllers/flatpickr_controller.js:88]: This `or` wording seems to be unnecessary, you're absolutely right:

##### with
Capture d’écran 2022-09-05 à 15 17 55

##### without
Capture d’écran 2022-09-05 à 15 17 20

I'd vote to remove it, see commit 3a72c4e716

File: app/webpacker/controllers/flatpickr_controller.js (hunk 4 above)
Comment (jibees) [reply app/webpacker/controllers/flatpickr_controller.js:15]: Maybe worth a new issue/PR? Because it's not that easy here (dynamic import, list of available languages, ...)

**Review by mkllnk — APPROVED** [review]
You are right, this pull request is big enough. Let's do the languages in a separate task.

**Comment by drummer83** [comment]
Hi @jibees,
Being a user of OFN I have been waiting for this one! :-) So thanks for working on it!

I have tested the following pages and found two issues which should be fixed, I think.

Note: All of this has only been tested as super-admin for now.

| Page | Date Range Picker | Status |
|-|-|-|
| orders | yes | OK |
| bulk order management | yes | can lead to loading forever |
| order cycles | no | OK |
| subscriptions | no | no date picker for end date |
| reports | no | OK |

## Issues
### Bulk Order Management: Query taking forever
There is a first query sent as soon as the start date has been selected. If I want to find an order some time long in the past all orders from then until now will be filtered - which takes forever and can't be stopped. See screen capture below:
Peek 2022-09-08 14-54
Proposals:
a) Start the query only after selecting start AND end date (not after start date already).
b) Do not start the query automatically but use the 'Filter results' and 'Clear filters' button like on orders page.
image

### Subscriptions: No date picker for Subscription end date
There used to be a date picker for the start date as well as the end date for subscriptions. Now the date picker for the end date is not showing, end date has to be typed manually.
This applies to editing an existing subscription as well as creating a new one.
Peek 2022-09-08 15-05

Moving back to In Dev.
Maybe @openfoodfoundation/train-drivers-product-owners want to share their opinion about the proposals to solve the issue around Bulk Order Management.

**Comment by jibees** [comment]
Thanks a lot for catching those issues! 🙏 

> Given the slowness of BOM, perhaps a button is the way to go? Therefore the query is launch only when all filters are selected?

I'd say yes, but I'm afraid this is maybe out of scope of this PR. Maybe a compromise could be choose the solution a for now (ie. only search for orders when range is completly selected), and open an issue to add a "Search/Filter/Whatever" button?

**Comment by RachL** [comment]
Fine by me :)

**Comment by jibees** [comment]
Small modifications, moving into Test Ready then.

**Comment by jibees** [comment]
specs fails, and pass locally. Moving to In Dev to debug...

**Comment by jibees** [comment]
Only one _flaky_ spec, back in `Code Review` then.

**Comment by drummer83** [comment]
Ok, thanks for checking again @jibees!
Here we go again:

| Page | Date Range Picker | Status (super-admin) | Status (enterprise user) |
|-|-|-|-|
| orders | yes | OK |  |
| bulk order management | yes | End date from PREVIOUS selection seems to be used |
| order cycles | no | OK |  |
| subscriptions | no | OK |  |
| reports | no | OK |  |

### Bulk order management
Something is wrong with the end date here. It seems like the end date from the previous search is being used. This may only occur when changing the month.
You can see in the screen capture, that there are many orders from September shown even though I selected a few days in August only.
Also, results with empty 'completed at' date are sometimes shown, sometimes they are not. This may be related.
I tried to check the queries - I don't understand much but I could definitely see previous dates there...

Peek 2022-09-12 16-32

We need to have another look. Back to In Dev.

**Comment by jibees** [comment]
@drummer83 
Thanks a lot for this one. This is strange, but let's tackle this one! 
I've tested locally, and I could not reproduce.
Maybe you can send me the request sent to do the search. URL path should be `/api/v0/orders.json` and payload should look like:
```
q[completed_at_gteq]: 2022-08-02T00:00:00+02:00
q[completed_at_lt]: 2022-08-06T00:00:00+02:00
q[completed_at_not_null]: true
q[distributor_id_eq]: 
q[order_cycle_id_eq]: 
q[shipment_state_not_eq]: shipped
q[state_not_eq]: canceled
```

Thanks!

**Comment by drummer83** [comment]
Ok @jibees, to reproduce:
1. Refresh the BOM page (the last 7 days are selected as date range automatically)
2. use date ranger picker to move to August and select 1. - 4. August
3. See results and query

Peek 2022-09-12 17-09

Here are my queries from the example above:

```
https://staging.openfoodnetwork.org.uk/admin/bulk_line_items.json? q%5Border_completed_at_gteq%5D=2022-08-01T00:00:00%2B02:00&q%5Border_completed_at_lt%5D=2022-09-13T00:00:00%2B02:00&
q%5Border_completed_at_not_null%5D=true&
q%5Border_distributor_id_eq%5D=&
q%5Border_order_cycle_id_eq%5D=&
q%5Border_shipment_state_not_eq%5D=shipped&
q%5Border_state_not_eq%5D=canceled&
q%5Bvariant_product_supplier_id_eq%5D=

https://staging.openfoodnetwork.org.uk/api/v0/orders.json?
q%5Bcompleted_at_gteq%5D=2022-08-01T00:00:00%2B02:00&q%5Bcompleted_at_lt%5D=2022-09-13T00:00:00%2B02:00&q%5Bcompleted_at_not_null%5D=true&q%5Bdistributor_id_eq%5D=&q%5Border_cycle_id_eq%5D=&q%5Bshipment_state_not_eq%5D=shipped&q%5Bstate_not_eq%5D=canceled

https://staging.openfoodnetwork.org.uk/api/v0/orders.json?q%5Bcompleted_at_gteq%5D=2022-08-01T00:00:00%2B02:00&q%5Bcompleted_at_lt%5D=2022-08-05T00:00:00%2B02:00&q%5Bcompleted_at_not_null%5D=true&q%5Bdistributor_id_eq%5D=&q%5Border_cycle_id_eq%5D=&q%5Bshipment_state_not_eq%5D=shipped&q%5Bstate_not_eq%5D=canceled

https://staging.openfoodnetwork.org.uk/admin/bulk_line_items.json?q%5Border_completed_at_gteq%5D=2022-08-01T00:00:00%2B02:00&q%5Border_completed_at_lt%5D=2022-08-05T00:00:00%2B02:00&q%5Border_completed_at_not_null%5D=true&q%5Border_distributor_id_eq%5D=&q%5Border_order_cycle_id_eq%5D=&q%5Border_shipment_state_not_eq%5D=shipped&q%5Border_state_not_eq%5D=canceled&q%5Bvariant_product_supplier_id_eq%5D=
```

**Comment by drummer83** [comment]
Probably still staged on UK. I used super admin access.

**Comment by jibees** [comment]
Super strange, and I now I can see why I can't reproduce it in my local environnement. 

Actually, when selecting the August 5, we send two requests: one from August from 1 to 5, and the other from August 1 to September 14. As the second one is longer to process and then receive, this one is used to displayed the results. 

Now I need to understand why two requests are sent!

**Comment by jibees** [comment]
@filipefurtad0 

I'm stuck on this on, and I probably need some help.
Spec `spec/system/admin/bulk_order_management_spec.rb:572` actually fail on line `#578` because it does not load the orders.

Actually, when trying to reproduce the bug on my own, it's work as expected: orders are loaded when selecting the range (ie. the start date and the end date).

When reverting the changes made on commit e7249599f1 on file `app/views/spree/admin/orders/bulk_management.html.haml`, it "works" (but actually introduce a new bug which is that the request is sent twice, one for the start date event, the other for the end date event).  

If you have time to find anything... I'll appreciate a lot ;)

**Comment by filipefurtad0** [comment]
Hey @jibees ,

> Spec spec/system/admin/bulk_order_management_spec.rb:572 actually fail on line #578 because it does not load the orders.

Indeed this was failing for me as well -> in master (!)

This was happening because the spec does not control the time zone of the test, which means sometimes it just adds the wrong dates on the calendar. By default the zone is 'Melbourne'. After I set the local time-zone to my real timezone there was no discrepancy between the dates set for the orders and those in the calendars.

So, in **master**, after I've added the line `TIMEZONE="Lisbon"` to `openfoodnetwork/.env.test` the test **passes**.

(Another option is to define a configuration option for the spec itself. I found [this to be an interesting suggestion](https://thiagoborg.es/blog/2018/03/18/timezones-for-rspec/) and it worked well on this spec.)

After doing this I've pulled your branch and installed `npm i stimulus-flatpickr` (somehow yarn install did not work...) and ran the spec. It did not fail on my system - and again, because I did not control the the timezone.

So, on **this branch** and on commit 3d95783786a126e22685498ff295aa9fee366ffd after I've added the line `TIMEZONE="Lisbon"` to `openfoodnetwork/.env.test` the test **fails**.

I'll need to investigate a bit more and look at the screenshots maybe. Let me know your thoughts.

**Comment by jibees** [comment]
Aside timezones issues,  It seems that when selecting date on line `#577` via `select_dates_from_daterangepicker(from, today)` no request to retrieve orders are sent. I need to understand why...

**Review by mkllnk — APPROVED** [review]
Nice. :sunglasses:

**Comment by drummer83** [comment]
Hi @jibees,
I had another look and I still see the problem on bulk order management page.

You can see in the screenshot below:
- Selected date is 02. August 2022 to 05. August 2022
- There are two orders.json? requests sent:
  - from selected start date (02.08.) until previous end date (05.10. - which was selected automatically when loading the page)
  - from selected start date (02.08.) until selected end date (05.08. 23:59:59 or 06.08. 00:00:00)
- Results are shown for the first request
image

Changing the date again shows the same behavior:
- Two requests are sent:
  - from selected start date (04.07.) to previous end date (06.08.)
  - from selected start date (04.07.) to selected end date (08.07.)
image

Sorry, I have to move to In Dev again.

**Comment by jibees** [comment]
@drummer83 
Thanks a sorry, this is a bit difficult actually ;)
I've made a modification, let's see if it's enough ... Thanks! 🙏

**Comment by drummer83** [comment]
Ok, thanks for checking again @jibees!
Here we go again:

| Page | Date Range Picker | Status (super-admin) | Status (enterprise user) |
|-|-|-|-|
| orders | yes | OK | OK |
| bulk order management | yes | OK | OK |
| edit order cycles | no | OK | OK |
| create new order cycles | no | OK | OK |
| edit subscriptions | no | OK | OK |
| create new subscriptions | no | OK | OK |
| reports | no | OK | OK |

Awesome, now you fixed it!!! :tada: 

Here you can see, that there is only one query sent (red) for the correct dates. Dates in yellow are the standard dates when loading the page.
image

This is ready to go now! Cheers! :beers: 
Thanks!

**Comment by mkllnk** [comment]
This introduced a regression which was hidden in our specs due to flakiness and retries:

* https://github.com/openfoodfoundation/openfoodnetwork/issues/9757
